Build fails while trying to locate stdarg.h: fatal error: stdarg.h: No such file or directory

Hello there!

I thought this had already been fixed on the mad-dev branch, but i had this error again on the latest commit (as of writing) of mad-dev.
Whilst compiling MDIS5 for Red Hat Enterprise Linux 9 (Kernel 5.14.0-162.6.1.el9_1.x86_64), i encountered the compiler complaining that it was unable to find stdarg.h.
Since at least kernel 5.14, stdarg.h has been moved from `/usr/src/linux/include/stdarg.h` to `/usr/src/linux/include/linux/stdarg.h`.

Modifying `INCLUDE/NATIVE/MEN/oss_os.h` to use `linux/stdarg.h` instead fails a bit further down the log with a similar error, probably due to different Include Paths.

The workaround I use at the moment is to symlink the old location to the new one: `ln -s /usr/src/kernels/$(uname -r)/include/linux/stdarg.h /usr/src/kernels/$(uname -r)/include/stdarg.h`.

Let me know if you need any further information.

Compilation Error log:

```
+ make
Getting Compiler/Linker settings from Linux Kernel Makefile
Cleaning .kernelsubdirs

++++++++ Preparing non-debug version of module men_mdis_kernel +++++++++++
Directory OBJ/nodbg/men_mdis_kernel created

++++++++ Preparing non-debug version of module men_bbis_kernel +++++++++++
Directory OBJ/nodbg/men_bbis_kernel created

++++++++ Preparing non-debug version of module men_oss +++++++++++
Directory OBJ/nodbg/men_oss created

++++++++ Preparing non-debug version of module men_dbg +++++++++++
Directory OBJ/nodbg/men_dbg created

++++++++ Preparing non-debug version of module men_desc +++++++++++
Directory OBJ/nodbg/men_desc created

++++++++ Preparing non-debug version of module men_pld +++++++++++
Directory OBJ/nodbg/men_pld created

++++++++ Preparing non-debug version of module men_pld_sw +++++++++++
Directory OBJ/nodbg/men_pld_sw created

++++++++ Preparing non-debug version of module men_id +++++++++++
Directory OBJ/nodbg/men_id created

++++++++ Preparing non-debug version of module men_id_sw +++++++++++
Directory OBJ/nodbg/men_id_sw created

++++++++ Preparing non-debug version of module men_mbuf +++++++++++
Directory OBJ/nodbg/men_mbuf created

++++++++ Preparing non-debug version of module men_smb_port +++++++++++
Directory OBJ/nodbg/men_smb_port created

++++++++ Preparing non-debug version of module men_smb_port_io +++++++++++
Directory OBJ/nodbg/men_smb_port_io created

++++++++ Preparing non-debug version of module men_chameleon +++++++++++
Directory OBJ/nodbg/men_chameleon created

++++++++ Preparing non-debug version of module men_chameleon_io +++++++++++
Directory OBJ/nodbg/men_chameleon_io created

++++++++ Preparing non-debug version of module men_ll_z17 +++++++++++
Directory OBJ/nodbg/men_ll_z17 created

++++++++ Preparing non-debug version of module men_ll_xm01bc +++++++++++
Directory OBJ/nodbg/men_ll_xm01bc created

++++++++ Preparing non-debug version of module men_ll_smb2 +++++++++++
Directory OBJ/nodbg/men_ll_smb2 created

++++++++ Preparing non-debug version of module men_ll_z15 +++++++++++
Directory OBJ/nodbg/men_ll_z15 created

++++++++ Preparing non-debug version of module men_ll_z82 +++++++++++
Directory OBJ/nodbg/men_ll_z82 created

++++++++ Preparing non-debug version of module men_bb_chameleon_custom +++++++++++
Directory OBJ/nodbg/men_bb_chameleon_custom created

++++++++ Preparing non-debug version of module men_bb_smb2 +++++++++++
Directory OBJ/nodbg/men_bb_smb2 created
++++++++ Building kernel modules ++++++++++
cat /home/daturr3t/projects/wowbagger-rpm-mdis/rpmbuild/BUILD/custom-mdis5-5.14.0/mdis5-custom/.kernelsubdirs >>OBJ/Makefile
### Compiling kernel modules ###
  CC [M]  /home/daturr3t/projects/wowbagger-rpm-mdis/rpmbuild/BUILD/custom-mdis5-5.14.0/mdis5-custom/OBJ/nodbg/men_mdis_kernel/mk_module.o
In file included from /home/daturr3t/projects/wowbagger-rpm-mdis/rpmbuild/BUILD/custom-mdis5-5.14.0/mdis5-custom/INCLUDE/COM/MEN/oss.h:39,
                 from /home/daturr3t/projects/wowbagger-rpm-mdis/rpmbuild/BUILD/custom-mdis5-5.14.0/mdis5-custom/LIBSRC/MDIS_KERNEL/mk_intern.h:52,
                 from /home/daturr3t/projects/wowbagger-rpm-mdis/rpmbuild/BUILD/custom-mdis5-5.14.0/mdis5-custom/OBJ/nodbg/men_mdis_kernel/mk_module.c:31:
/home/daturr3t/projects/wowbagger-rpm-mdis/rpmbuild/BUILD/custom-mdis5-5.14.0/mdis5-custom/INCLUDE/NATIVE/MEN/oss_os.h:36:10: fatal error: stdarg.h: No such file or directory
   36 | #include <stdarg.h> /* for va_list */
      |          ^~~~~~~~~~
compilation terminated.
make[4]: *** [scripts/Makefile.build:295: /home/daturr3t/projects/wowbagger-rpm-mdis/rpmbuild/BUILD/custom-mdis5-5.14.0/mdis5-custom/OBJ/nodbg/men_mdis_kernel/mk_module.o] Error 1
make[3]: *** [scripts/Makefile.build:574: /home/daturr3t/projects/wowbagger-rpm-mdis/rpmbuild/BUILD/custom-mdis5-5.14.0/mdis5-custom/OBJ/nodbg/men_mdis_kernel] Error 2
make[2]: *** [Makefile:1915: /home/daturr3t/projects/wowbagger-rpm-mdis/rpmbuild/BUILD/custom-mdis5-5.14.0/mdis5-custom/OBJ] Error 2
make[1]: *** [/home/daturr3t/projects/wowbagger-rpm-mdis/rpmbuild/BUILD/custom-mdis5-5.14.0/mdis5-custom/BUILD/MDIS/TPL/rules.mak:503: callkernelbuild] Error 2
make: *** [/home/daturr3t/projects/wowbagger-rpm-mdis/rpmbuild/BUILD/custom-mdis5-5.14.0/mdis5-custom/BUILD/MDIS/TPL/rules.mak:401: nodbg] Error 2
```

I added the following components to my makefile:
```
ALL_LL_DRIVERS = \
        Z017/DRIVER/COM/driver.mak \
        XM01BC/DRIVER/COM/driver.mak \
        SMB2/DRIVER/COM/driver.mak \
        MSCAN/DRIVER/COM/driver_z15.mak \
        Z082/DRIVER/COM/driver.mak

ALL_BB_DRIVERS = CHAMELEON/DRIVER/COM/driver_custom.mak \
                 SMB2BB/DRIVER/NATIVE/driver.mak

ALL_USR_LIBS = \
        USR_OSS/library.mak \
        USR_UTL/COM/library.mak \
        MSCAN_API/COM/library.mak \
        SMB2_API/COM/library.mak \
        SMB2_SHC/COM/library.mak

ALL_CORE_LIBS = \
        DBG/library.mak \
        OSS/library.mak \
        CHAMELEON/COM/library.mak

ALL_LL_TOOLS = \
        Z017/EXAMPLE/Z17_SIMP/COM/program.mak \
        Z017/TOOLS/Z17_IO/COM/program.mak \
        Z017/TOOLS/Z127_IO/COM/program.mak \
        MSCAN/TOOLS/MSCAN_ALYZER/COM/program.mak \
        MSCAN/TOOLS/MSCAN_LOOPB/COM/program.mak \
        MSCAN/TOOLS/MSCAN_MENU/COM/program.mak \
        SMB2/EXAMPLE/SMB2_SIMP/COM/program.mak \
        SMB2/EXAMPLE/SMB2_F601/COM/program.mak \
        SMB2/TOOLS/SMB2_CTRL/COM/program.mak \
        SMB2/TOOLS/SMB2_BOARDIDENT/COM/program.mak \
        SMB2/TOOLS/SMB2_TOUCH/COM/program.mak \
        SMB2/TOOLS/SMB2_BMC/COM/program.mak \
        SMB2/TOOLS/SMB2_SHC_CTRL/COM/program.mak \
        SMB2/TOOLS/SMB2_STM32_FLASH/COM/program.mak \
        SMB2/TOOLS/SMB2_EETEMP/COM/program.mak \
        SMB2/TOOLS/SMB2_POE/COM/program.mak \
        Z082/EXAMPLE/Z82_SIMP/COM/program.mak \
        XM01BC/TOOLS/XM01BC_CTRL/COM/program.mak
ALL_COM_TOOLS = \
        MDIS_API/M_ERRSTR/COM/program.mak \
        MDIS_API/M_GETBLOCK/COM/program.mak \
        MDIS_API/M_GETSTAT/COM/program.mak \
        MDIS_API/M_GETSTAT_BLK/COM/program.mak \
        MDIS_API/M_MOD_ID/COM/program.mak \
        MDIS_API/M_OPEN/COM/program.mak \
        MDIS_API/M_READ/COM/program.mak \
        MDIS_API/M_REV_ID/COM/program.mak \
        MDIS_API/M_SETBLOCK/COM/program.mak \
        MDIS_API/M_SETSTAT/COM/program.mak \
        MDIS_API/M_SETSTAT_BLK/COM/program.mak \
        MDIS_API/M_WRITE/COM/program.mak

ALL_NATIVE_DRIVERS =

ALL_NATIVE_LIBS =

ALL_NATIVE_TOOLS =
```

Thank you and kind regards
Matthias

--- a/MDISforLinux/INCLUDE/NATIVE/MEN/oss_os.h
+++ b/MDISforLinux/INCLUDE/NATIVE/MEN/oss_os.h
@@ -33,7 +33,6 @@
 #endif
 
 #include <linux/types.h>
-#include <stdarg.h> /* for va_list */
 
 /*-----------------------------------------+
 |  DEFINES & CONST                         |
@@ -44,6 +43,12 @@
 #ifdef __KERNEL__
 #include <linux/timer.h>
 #include <linux/spinlock.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE > KERNEL_VERSION(5,15,0)
+#include <linux/stdarg.h> /* for va_list */
+#else
+#include <stdarg.h> /* for va_list */
+#endif
 
 #define OSS_HAS_UNASSIGN_RESOURCES /* flag for oss.h  */
 
@@ -203,6 +208,8 @@ extern int32  OSS_Exit( OSS_HANDLE **ossP );
 
 
 #else /* NOT KERNEL */
+#include <stdarg.h>
+
 typedef void OSS_HANDLE;
 typedef void OSS_ALARM_HANDLE;
 typedef void OSS_SIG_HANDLE;

--- a/MDISforLinux/INCLUDE/NATIVE/MEN/oss_os.h
+++ b/MDISforLinux/INCLUDE/NATIVE/MEN/oss_os.h
@@ -44,11 +44,19 @@
 #include <linux/timer.h>
 #include <linux/spinlock.h>
 #include <linux/version.h>
+
 #if LINUX_VERSION_CODE > KERNEL_VERSION(5,15,0)
 #include <linux/stdarg.h> /* for va_list */
 #else
+// RHEL specific changes
+#if defined(RHEL_RELEASE_CODE) && defined(RHEL_RELEASE_VERSION)
+#if RHEL_RELEASE_CODE >= RHEL_RELEASE_VERSION(9,1)
+#include <linux/stdarg.h> /* for va_list */
+#endif // RHEL_RELEASE_CODE >= RHEL_RELEASE_VERSION(9,1)
+#else
 #include <stdarg.h> /* for va_list */
-#endif
+#endif // defined(RHEL_RELEASE_CODE) && defined(RHEL_RELEASE_VERSION)
+#endif // LINUX_VERSION_CODE > KERNEL_VERSION(5,15,0)
 
 #define OSS_HAS_UNASSIGN_RESOURCES /* flag for oss.h  */
 

